app/importers/file_importers/base_csv_importer.rb

```diff
@@ -0,0 +1,36 @@
+class BaseCsvImporter < Struct.new :school_scope, :client, :log, :progress_bar, :import_record_detail
```

Comment:
Question 2: This looks like you've awesomely factored some bits around processing CSV files into what's essentially a concrete class that does a clear, well-defined thing.  But here it's implementing as an abstract base class that's intended to be subclassed, and that adds some more scope creep.

In particular, I'm noticing the `filter` and `before_import` and `after_import` hooks.  Adding hooks like that are really good if you have to give up control of what's happening to another class.  But for most code, unless you really want to describe things declaratively and commit to those abstractions, adding hooks like this creates  barriers to change.  Here we're taking a step towards essentially creating a small DSL for programming primitives (filter, before, after), which is usually a smell that we don't have the abstraction we need - if we did we'd just write these things in plain Ruby :)

So I'm wondering about what it would look like if we made this a concrete class.  `CsvProcessor` might take a bunch of parameters `(school_scope, client, log, progress_bar, import_record_detail, remote_file_name, data_transformer)`.  It has some defaults, and callers can pass parameters instead of subclassing and overriding methods.  It's single method, `#process` could take a block, which it then yields to for each transformed row.  The caller can do whatever it wants with it with the row, filtering, etc.  With a concrete class like that, individual importer classes can call it in their `#import` methods.  That removes the indirection with the hooks.  Sketching this out, it might look something like this:

```
class SectionsImporter
  def import
    csv_processor = CsvProcessor.new({ remote_filename: 'sections.csv', ... })
    # do something before like initialize a set of records we're tracking
    csv_processor.process {|row| process_row(row) }
    # do something after, like delete stale rows
  end

  private
  # This receives the transformed row, and then we do whatever you want with it, skip it, update, import
  def process_row(row, import_record_detail)
    # should it skip it?
    # import it and log
    nil
  end
end
```

The benefit being that we've removed the concepts around filter, before and after (since we can say them in plain Ruby now), and the control flow is more explicit within each importer.

Reply:
@kevinrobinson You're correct, I was creating what is essentially an abstract class (needs to be subclassed, would throw an error in compiled languages if you try to instantiate it). Typically abstract classes do provide some functionality, otherwise you would use an interface. 

Of course, none of this is very rubyish. I'm likely trying to solve a problem that doesn't yet exist, namely importing a wider variety of files for different districts.

So I think for now, it's best to leave things as they are and add some good logging without changing the object structure of the importers.

Reply:
@jhilde Yeah I think I'd have the same suggestion in Java too - using collaborator objects is another way to share functionality too, and usually has benefits over abstract base classes.  By providing an abstract base class as an API, you're really committing to an abstraction and callers are forced to give up control to the base class, who "owns" the flow of what is happening.  The subclass can only answer specific questions that the abstract base class anticipated in advance, it can't change how it uses the base class' functionality, or compose it in new ways.  And changing the base class becomes increasingly hard, since you can't breaking it apart and recompose it as easily as you can when composing via collaborators and normal method calls.

If you're aiming for reducing duplication, you can often do that more directly by factoring the duplication out into a concrete class that you then use in a few different places with plain method calls.  That makes the control flow more explicit - callers are always in control except for when they call methods on collaborators.  This is most important when you want to preserve flexibility and optionality to compose those pieces in different ways in the future - if you have a few small pieces with clear lines between them, it's easier to build new things from it versus a fatter base class that does all the things.

Anyway, I think it's awesome you're looking at this and at ways to improve factoring!  👍 But agreed that it drifts a bit from improving logging.